### ISSUE TITLE
Add EWKB encoding and decoding support on geobase

### ISSUE THREAD

**Issue by navispatial** [navibyte/geospatial#165]
See also #29 and #224.

*2024-04-09 .. 2024-04-16: Updated plan for this issue*

As noted in comments support for decoding EWKB data (however ignoring any SRID) is added to [geobase version 1.0.2](https://pub.dev/packages/geobase/versions/1.0.2).

Full support is targeted on v1.1.0.

At least following features should be implemented for full EWKB support:
* Currently there is the [WKB](https://pub.dev/documentation/geobase/latest/geobase/WKB-class.html) class provides both encoder and decoder for WKB
  * Decoding implementation (already updated for v1.0.2) would be shared for both WKB and EWKB
  * Encoding implementation must be selected by an used, either `WkbFlavor.standard` or `WkbFlavor.extented`.
*  separate utility function just to get metadata (like SRID) from geometry data like EWKB bytes or hex strings (relatively small change as other classes do not need to be changed)
* Helpers to decode directly from hex strings
  * Currently geometry classes have factories like `Point.decode()` that take `Uint8List` byte data as input
  * There could be also factories like `Point.decodeHex()` that take string with hex data as input
* Sample code and documentation

Implementation for v1.1.0:
* Default endianess to Endian.little in WKB and some docs updated. 
* Support EWKB dimensionality flags for z and m.
* CoordRefSys to support getting EPSG code as an integer.
* Support encoding SRID to EWKB binary data.
* Utilities to decode a byte order, a WKB flavor, a coordinate type and SRID from WKB bytes.
* A sample for decoding and encoding EWKB binary data.

**Comment by navispatial** [comment]
A question how to get geographic longitude and latitude values from geographic point generated by PostGIS, like `0101000020E6100000B8A1164EC5725140D1650CC83FB83840`.

This hex string seems to be EWKB encoded string. This encoding is not currently supported by the `geobase` package (and not by `geocore` package either).

Below is some experimenting how to decode longitude and latitude from EWKB string using the WKB support (decode / encode) provided by the `geobase` package.

```dart
import 'dart:typed_data';

import 'package:convert/convert.dart';
import 'package:geobase/vector.dart';
import 'package:geobase/vector_data.dart';

/// This sample decodes geographic longitude/latitude from EKWB endoded text.
void main() {
  // PostGIS uses EWKB encoding
  // See => https://postgis.net/docs/using_postgis_dbmanagement.html#EWKB_EWKT
  //
  // Encoding for a 2D point:
  // - 1 byte: endianess (uint 8)
  // - 4 bytes: geometry type (uint 32)
  // - 4 bytes: SRD id (uint 32)
  // - 8 bytes: longitude (float 64)
  // - 8 bytes: latitude (float 64)
  // 
  // These are 2D point samples encoded as EWKB in hex string.
  const hexEWKBsamples = [
    '01010000200400000000000000000000000000000000000000',
    '0101000020AB1000000000000000C05EC00000000000004140',
    '0101000020E6100000B8A1164EC5725140D1650CC83FB83840',
  ];

  // The geobase package supports only WKB encoding, for example a 2D point:
  // - 1 byte: endianess (uint 8)
  // - 4 bytes: geometry type (uint 32)
  // - 8 bytes: longitude (float 64)
  // - 8 bytes: latitude (float 64)
  // Also there are some differences on geometry types.
  //
  // See WKB:
  // https://en.wikipedia.org/wiki/Well-known_text_representation_of_geometry
  //
  // So first convert from EWKB to WKB, and then decode geographic coordinates.
  for (final ewkbHex in hexEWKBsamples) {
    // EWKB 2D point to WKB 2D point ("fixes" geom type and omits SRS id)
    // (note this is not a general purpose conversion for all geom types) 
    final wkbHex = '${ewkbHex.substring(0, 8)}00${ewkbHex.substring(18)}';
  
    // convert WKB hex representation to bytes
    final wkbBytes = Uint8List.fromList(hex.decode(wkbHex));

    // decode bytes to Point geometry and get position from it as geographic
    final point =
        Point.decode(wkbBytes, format: WKB.geometry).position.asGeographic;

    // print longitude, latitude
    print('EWKB($ewkbHex)');
    print('  => (lon, lat) ${point.lon} ${point.lat}');

    // This should print:
    // EWKB(01010000200400000000000000000000000000000000000000)
    //   => (lon, lat) 0.0 0.0
    // EWKB(0101000020AB1000000000000000C05EC00000000000004140)
    //   => (lon, lat) -123.0 34.0
    // EWKB(0101000020E6100000B8A1164EC5725140D1650CC83FB83840)
    //   => (lon, lat) 69.79329254350785 24.719723227526462
  }
}
```

**Comment by ethicnology** [comment]
Dude I was struggling with the EWKB format to decode geographic postgis point, and this snipper helped a lot, thank you.

I hope your library will implement this

**Comment by navispatial** [comment]
Implemented in [geobase version 1.1.0](https://pub.dev/packages/geobase/versions/1.1.0), see [milestone](https://github.com/navibyte/geospatial/milestone/1).

See also the article [Decode and encode GeoJSON, WKT and WKB in Dart and Flutter apps](https://medium.com/@navibyte/decode-and-encode-geojson-wkt-and-wkb-in-dart-and-flutter-apps-ab2ef4ece2f1)
